diesel/src/query_builder/select_statement/select_by/mod.rs

```diff
@@ -0,0 +1,137 @@
+/// `Query`/`AsQuery`, `AppendSelection` and `IntoUpdateTarget`.
+/// note: it would have conflicting implementation of `LoadQuery` if
+/// this implements `AsQuery`.
+/// It does not support `GroupByDsl` right now (using these dsl would
+/// unboxing into inner statement).
+/// When using `SelectDsl`/`SelectByDsl` would intendedly unbox/rebox
+/// Other operation should keep Selection invariant.
+#[doc(hidden)]
+#[must_use = "Queries are only executed when calling `load`, `get_result` or similar."]
+pub struct SelectByStatement<Selection, Statement> {
+    pub(crate) select: PhantomData<Selection>,
+    pub(crate) inner: Statement,
+}
+
+impl<S, Stmt> fmt::Debug for SelectByStatement<S, Stmt>
+where
+    Stmt: fmt::Debug,
+{
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("SelectByStatement")
+            .field("inner", &self.inner)
+            .finish()
+    }
+}
+
+impl<S, Stmt> Clone for SelectByStatement<S, Stmt>
+where
+    Stmt: Clone,
+{
+    fn clone(&self) -> Self {
+        SelectByStatement::new(self.inner.clone())
+    }
+}
+
+impl<S, Stmt> Copy for SelectByStatement<S, Stmt> where Stmt: Copy {}
+
+impl<S, Stmt> QueryId for SelectByStatement<S, Stmt>
+where
+    Stmt: QueryId,
+{
+    type QueryId = Stmt::QueryId;
+    const HAS_STATIC_QUERY_ID: bool = Stmt::HAS_STATIC_QUERY_ID;
+}
+
+impl<S, ST> SelectByStatement<S, ST> {
+    pub(crate) fn new(inner: ST) -> Self {
+        SelectByStatement {
+            inner,
+            select: Default::default(),
+        }
+    }
+}
+
+impl<DB, S, Stmt> QueryFragment<DB> for SelectByStatement<S, Stmt>
+where
+    DB: Backend,
+    Stmt: QueryFragment<DB>,
+{
+    fn walk_ast(&self, mut out: AstPass<DB>) -> QueryResult<()> {
+        self.inner.walk_ast(out.reborrow())?;
+        Ok(())
+    }
+}
+
+impl<ST, S, Stmt> SelectQuery for SelectByStatement<S, Stmt>
+where
+    Stmt: SelectQuery<SqlType = ST>,
+{
+    type SqlType = ST;
+}
+
+impl<S, Stmt> SelectByQuery for SelectByStatement<S, Stmt>
+where
+    S: Selectable,
+    Stmt: SelectByQuery<Expression = S::Expression>,
+{
+    type Expression = S::Expression;
+}
+
+impl<S, Stmt, QS> ValidSubselect<QS> for SelectByStatement<S, Stmt>
+where
+    Self: SelectQuery,
+    Stmt: ValidSubselect<QS>,
+{
+}
+
+impl<S, Stmt, T> AppearsInFromClause<T> for SelectByStatement<S, Stmt>
+where
+    Stmt: AppearsInFromClause<T>,
+{
+    type Count = Stmt::Count;
+}
+
+impl<S, Stmt> QuerySource for SelectByStatement<S, Stmt>
+where
+    S: Selectable,
+    Stmt: QuerySource,
+    S::Expression: SelectableExpression<Self>,
+{
+    type FromClause = Stmt::FromClause;
+    type DefaultSelection = S::Expression;
+
+    fn from_clause(&self) -> Self::FromClause {
+        self.inner.from_clause()
+    }
+
+    fn default_selection(&self) -> Self::DefaultSelection {
+        S::new_expression()
+    }
+}
+
+impl<Conn, S, Stmt> LoadQuery<Conn, S> for SelectByStatement<S, Stmt>
```

Comment:
I'm not sure if it would be meaningful to also implement `ExecuteDsl`?

Reply:
I'm not sure but I think now `SelectByStatement` implements `QueryFragment` and  `QueryId`, it would automatically implement `ExecuteDsl`
https://github.com/diesel-rs/diesel/blob/bebdce636532465f2d6f6d0a2a0d5913f6d7b0d3/diesel/src/query_dsl/load_dsl.rs#L48-L57

Reply:
That seems to be correct :+1: